Monitor interface changes using nw_monitor_path_t

youennf opened the pull request:
#### 6c032159699e2a5fcd3a8da54e3ad0f1a8e4c35d
<pre>
Monitor interface changes using nw_monitor_path_t
<a href="https://rdar.apple.com/147548855">rdar://147548855</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=290165">https://bugs.webkit.org/show_bug.cgi?id=290165</a>

Reviewed by Chris Dumez and Alex Christensen.

Instead of a timer based approach, we use nw_monitor to react upon network changes.
We remove the use of nw_interface_create_with_name SPI since it is no longer needed.

* Source/WebCore/PAL/pal/spi/cocoa/NetworkSPI.h:
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCMonitor.cpp:
(WebKit::NetworkManager::NetworkManager):
(WebKit::NetworkManager::addListener):
(WebKit::NetworkManager::removeListener):
(WebKit::interfaceAdapterType):
(WebKit::gatherNetworkMap):
(WebKit::NetworkManager::adapterTypeFromInterfaceName const):
(WebKit::NetworkManager::updateNetworksFromPath):

Canonical link: <a href="https://commits.webkit.org/292604@main">https://commits.webkit.org/292604@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7f7a5dddcbfb08c64fa22ea726a83851c67bf32f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96523 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16137 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6229 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101595 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47043 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98568 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16433 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24571 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73580 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30808 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99526 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12364 "Found 1 new test failure: fast/forms/ios/focus-input-via-button.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87296 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53916 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12121 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5078 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46371 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82232 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5173 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103619 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23590 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17203 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82629 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23841 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83334 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82003 "Found 3 new API test failures: /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print-errors, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/close-after-print (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20583 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26649 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4172 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17064 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23553 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28708 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/23212 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26692 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24953 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/webrtc/NetworkRTCMonitor.cpp:
```diff
@@ -119,8 +128,18 @@ static NetworkManager& networkManager()
 NetworkManager::NetworkManager()
     : m_queue(ConcurrentWorkQueue::create("RTC Network Manager"_s))
+#if PLATFORM(COCOA)
+    , m_nwMonitor(WTF::adoptCF(nw_path_monitor_create()))
```
Do we really need `WTF::`?

youennf replied:
Will remove

cdumez reviewed (approved):
LGTM

achristensen07 commented on Source/WebKit/NetworkProcess/webrtc/NetworkRTCMonitor.cpp:
```diff
@@ -147,7 +169,11 @@ void NetworkManager::removeListener(NetworkRTCMonitor& monitor)
 
     RELEASE_LOG(WebRTC, "NetworkManagerWrapper stopUpdating");
+#if PLATFORM(COCOA)
+    nw_path_monitor_cancel(m_nwMonitor.get());
```
Can there ever be more than one listener?  Should we only cal start/cancel if we are adding the first or removing the last?

youennf replied:
There is one listener per web process that has instantiated a RTCPeerConnection.
Start/cancel is just an optimization

achristensen07 commented on Source/WebKit/NetworkProcess/webrtc/NetworkRTCMonitor.h:
```diff
@@ -31,6 +31,10 @@
 #include <wtf/WeakPtr.h>
 
+#if PLATFORM(COCOA)
+#include <Network/Network.h>
```
It seems like this could be in the cpp file.

youennf replied:
Yes, this was an unneeded leftover

youennf committed:
Monitor interface changes using nw_monitor_path_t
rdar://147548855
https://bugs.webkit.org/show_bug.cgi?id=290165

Reviewed by Chris Dumez and Alex Christensen.

Instead of a timer based approach, we use nw_monitor to react upon network changes.
We remove the use of nw_interface_create_with_name SPI since it is no longer needed.

* Source/WebCore/PAL/pal/spi/cocoa/NetworkSPI.h:
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCMonitor.cpp:
(WebKit::NetworkManager::NetworkManager):
(WebKit::NetworkManager::addListener):
(WebKit::NetworkManager::removeListener):
(WebKit::interfaceAdapterType):
(WebKit::gatherNetworkMap):
(WebKit::NetworkManager::adapterTypeFromInterfaceName const):
(WebKit::NetworkManager::updateNetworksFromPath):

Canonical link: https://commits.webkit.org/292604@main
